<a href="https://colab.research.google.com/github/myla48/titanic-survival-analysis/blob/main/Titanic_Part1_Part2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Part 1: Create  Dataset

In [6]:
import pandas as pd

# Part 1: Custom dataset
data = {
    "Name": ["Abel","Sara","John","Mimi","Kebede","Liya","Noah","Hana","Sam","Ruth","Ben","Sofi","Mark","Eden","Kal"],
    "Age": [20,21,19,22,23,20,21,22,23,24,25,20,19,18,21],
    "Grade": ["A","B","A","C","B","A","A","B","C","A","B","A","B","C","A"],
    "Gender": ["M","F","M","F","M","F","M","F","M","F","M","F","M","F","M"],
    "Score": [85,78,90,65,70,88,92,75,60,95,72,89,77,66,91]
}

index_labels = [f"S{i}" for i in range(1,16)]

df_custom = pd.DataFrame(data, index=index_labels)
df_custom  # Display Part 1 dataset


,Name,Age,Grade,Gender,Score
S1,Abel,20,A,M,85
S2,Sara,21,B,F,78
S3,John,19,A,M,90
S4,Mimi,22,C,F,65
S5,Kebede,23,B,M,70
S6,Liya,20,A,F,88
S7,Noah,21,A,M,92
S8,Hana,22,B,F,75
S9,Sam,23,C,M,60
S10,Ruth,24,A,F,95


---

## Part 2: Real Dataset — Titanic

### Step 1: Exploration

In [2]:
# Load the Titanic Dataset
titanic_df = pd.read_csv('/content/Titanic-Dataset.csv')

print("First 5 rows of the Titanic Dataset:")
display(titanic_df.head())

print("\nInformation about the Titanic Dataset:")
display(titanic_df.info())

print("\nDescriptive statistics of the Titanic Dataset:")
display(titanic_df.describe())

First 5 rows of the Titanic Dataset:


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S



Information about the Titanic Dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


None


Descriptive statistics of the Titanic Dataset:


,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### Step 2: Data Cleaning

In [3]:
# Handle missing values for 'Age' with the median
titanic_df['Age'].fillna(titanic_df['Age'].median(), inplace=True)

# Handle missing values for 'Embarked' with the mode
# Get the most frequent value in the 'Embarked' column
most_frequent_embarked = titanic_df['Embarked'].mode()[0]
titanic_df['Embarked'].fillna(most_frequent_embarked, inplace=True)

# Drop the 'Cabin' column due to a large number of missing values
titanic_df.drop('Cabin', axis=1, inplace=True)

# Remove duplicates if any (though unlikely in this dataset for passenger IDs)
initial_rows = titanic_df.shape[0]
titanic_df.drop_duplicates(inplace=True)
duplicates_removed = initial_rows - titanic_df.shape[0]

print(f"Missing values after cleaning (Age, Embarked):")
display(titanic_df[['Age', 'Embarked']].isnull().sum())
print(f"'Cabin' column dropped. Remaining columns: {list(titanic_df.columns)}")
print(f"Number of duplicate rows removed: {duplicates_removed}")

print("\nInformation about the Titanic Dataset after cleaning:")
display(titanic_df.info())

Missing values after cleaning (Age, Embarked):


/tmp/ipykernel_20521/2623726452.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  titanic_df['Age'].fillna(titanic_df['Age'].median(), inplace=True)
/tmp/ipykernel_20521/2623726452.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, 

,0
Age,0
Embarked,0


'Cabin' column dropped. Remaining columns: ['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Embarked']
Number of duplicate rows removed: 0

Information about the Titanic Dataset after cleaning:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


None

### Step 3: Data Analysis

In [4]:
# Survival rate by gender
survival_by_gender = titanic_df.groupby('Sex')['Survived'].mean() * 100
print("Survival Rate by Gender:")
display(survival_by_gender)

# Survival rate by class
survival_by_class = titanic_df.groupby('Pclass')['Survived'].mean() * 100
print("\nSurvival Rate by Class:")
display(survival_by_class)

# Average age per class
average_age_by_class = titanic_df.groupby('Pclass')['Age'].mean()
print("\nAverage Age per Class:")
display(average_age_by_class)

# Survival rate by age group
# Create age groups
titanic_df['AgeGroup'] = pd.cut(titanic_df['Age'], bins=[0, 12, 18, 60, 100], labels=['Child', 'Teenager', 'Adult', 'Senior'])
survival_by_age_group = titanic_df.groupby('AgeGroup')['Survived'].mean() * 100
print("\nSurvival Rate by Age Group:")
display(survival_by_age_group)

Survival Rate by Gender:


,Survived
Sex,
female,74.203822
male,18.890815



Survival Rate by Class:


,Survived
Pclass,
1,62.962963
2,47.282609
3,24.236253



Average Age per Class:


,Age
Pclass,
1,36.812130
2,29.765380
3,25.932627



Survival Rate by Age Group:


/tmp/ipykernel_20521/1659240725.py:19: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  survival_by_age_group = titanic_df.groupby('AgeGroup')['Survived'].mean() * 100


,Survived
AgeGroup,
Child,57.971014
Teenager,42.857143
Adult,36.575342
Senior,22.727273


### Step 4: Filtering

In [5]:
# Female passengers who survived
female_survivors = titanic_df[(titanic_df['Sex'] == 'female') & (titanic_df['Survived'] == 1)]
print("Female Passengers who Survived:")
display(female_survivors.head())
print(f"Total Female Survivors: {len(female_survivors)}")

# Children who survived (AgeGroup 'Child' and Survived == 1)
child_survivors = titanic_df[(titanic_df['AgeGroup'] == 'Child') & (titanic_df['Survived'] == 1)]
print("\nChildren who Survived:")
display(child_survivors.head())
print(f"Total Child Survivors: {len(child_survivors)}")

# Passengers in 1st class with high survival probability
# (This can be interpreted as 1st class passengers who survived)
first_class_survivors = titanic_df[(titanic_df['Pclass'] == 1) & (titanic_df['Survived'] == 1)]
print("\nPassengers in 1st Class who Survived:")
display(first_class_survivors.head())
print(f"Total 1st Class Survivors: {len(first_class_survivors)}")


Female Passengers who Survived:


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,AgeGroup
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,Adult
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,Adult
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,Adult
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,S,Adult
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,C,Teenager


Total Female Survivors: 233

Children who Survived:


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,AgeGroup
10,11,1,3,"Sandstrom, Miss. Marguerite Rut",female,4.00,1,1,PP 9549,16.7000,S,Child
43,44,1,2,"Laroche, Miss. Simonne Marie Anne Andree",female,3.00,1,2,SC/Paris 2123,41.5792,C,Child
58,59,1,2,"West, Miss. Constance Mirium",female,5.00,1,2,C.A. 34651,27.7500,S,Child
78,79,1,2,"Caldwell, Master. Alden Gates",male,0.83,0,2,248738,29.0000,S,Child
125,126,1,3,"Nicola-Yarred, Master. Elias",male,12.00,1,0,2651,11.2417,C,Child


Total Child Survivors: 40

Passengers in 1st Class who Survived:


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,AgeGroup
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,Adult
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,Adult
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,S,Adult
23,24,1,1,"Sloper, Mr. William Thompson",male,28.0,0,0,113788,35.5000,S,Adult
31,32,1,1,"Spencer, Mrs. William Augustus (Marie Eugenie)",female,28.0,1,0,PC 17569,146.5208,C,Adult


Total 1st Class Survivors: 136


### Step 5: Insights

Based on the analysis:

*   **Who was more likely to survive (gender)?**
    Women (`74.20%` survival rate) were significantly more likely to survive than men (`18.89%` survival rate).

*   **Did class affect survival?**
    Yes, class strongly affected survival. Passengers in 1st class had the highest survival rate (`62.96%`), followed by 2nd class (`47.28%`), and then 3rd class (`24.24%`). This indicates a clear advantage for higher classes.

*   **Were children prioritized?**
    Children (`57.97%` survival rate for 'Child' age group) had a higher survival rate compared to teenagers (`42.86%`), adults (`36.58%`), and seniors (`22.73%`). This suggests that children were indeed prioritized or had a better chance of survival.

*   **What combination had the highest survival rate?**
    Combining the insights, **female passengers in 1st class** would likely have had the highest survival rate. From the previous output, female survival was `74.20%` and 1st class survival was `62.96%`. While we didn't explicitly calculate the combined rate for 'female and 1st class' in a single line, these two groups individually show high survival, indicating a strong positive correlation for this combination.